Sources/ComposableArchitecture/SharedState/PersistenceKey/FileStorageKey.swift

```diff
@@ -0,0 +1,303 @@
+  /// interacts with file storage. For example, while your app is running for UI tests you
+  /// probably do not want your features writing changes to disk, which would cause that data to
+  /// bleed over from test to test.
+  ///
+  /// So, for that situation you can use the ``FileStorage/inMemory`` file storage so that each
+  /// run of the app starts with a fresh "file system" that will never interfer with other tests:
+  ///
+  /// ```swift
+  /// @main
+  /// struct EntryPoint: App {
+  ///   let store = Store(initialState: AppFeature.State()) {
+  ///     AppFeature()
+  ///   } withDependencies: {
+  ///     if ProcessInfo.processInfo.environment["UITesting"] == "true" {
+  ///       $0.defaultFileStorage = .inMemory
+  ///     }
+  ///   }
+  /// }
+  /// ```
+  public var defaultFileStorage: FileStorage {
+    get { self[FileStorageDependencyKey.self] }
+    set { self[FileStorageDependencyKey.self] = newValue }
+  }
+}
+
+/// A type that encapsulates saving and loading data from disk.
+public struct FileStorage: Sendable {
+  let id: AnyHashableSendable
+  let async: @Sendable (DispatchWorkItem) -> Void
+  let asyncAfter: @Sendable (DispatchTimeInterval, DispatchWorkItem) -> Void
+  let createDirectory: @Sendable (URL, Bool) throws -> Void
+  let fileExists: @Sendable (URL) -> Bool
+  let fileSystemSource:
+    @Sendable (URL, DispatchSource.FileSystemEvent, @escaping () -> Void) -> AnyCancellable
+  let load: @Sendable (URL) throws -> Data
+  @_spi(Internals) public let save: @Sendable (Data, URL) throws -> Void
+
+  /// File storage that interacts directly with the file system for saving, loading and listening
+  /// for file changes.
+  ///
+  /// This is the version of the ``Dependencies/DependencyValues/defaultFileStorage`` dependency
+  /// that is used by default when running your app in the simulator or on device.
+  public static var fileSystem = fileSystem(
+    queue: DispatchQueue(label: "co.pointfree.ComposableArchitecture.FileStorage")
+  )
+
+  /// File storage that emulates a file system without actually writing anything to disk.
+  ///
+  /// This is the version of the ``Dependencies/DependencyValues/defaultFileStorage`` dependency
+  /// that is used by default when running your app in tests and previews.
+  public static var inMemory: Self {
+    inMemory(fileSystem: LockIsolated([:]))
+  }
+
+  @_spi(Internals) public static func fileSystem(queue: DispatchQueue) -> Self {
+    Self(
+      id: AnyHashableSendable(queue),
+      async: { queue.async(execute: $0) },
+      asyncAfter: { queue.asyncAfter(deadline: .now() + $0, execute: $1) },
+      createDirectory: {
+        try FileManager.default.createDirectory(at: $0, withIntermediateDirectories: $1)
+      },
+      fileExists: { FileManager.default.fileExists(atPath: $0.path) },
+      fileSystemSource: {
+        let source = DispatchSource.makeFileSystemObjectSource(
+          fileDescriptor: open($0.path, O_EVTONLY),
```

Comment:
@Matejkob We'd take a PR that bails out on a negative file descriptor if you're down to submit! Since it's generally a developer-specified URL then failure to access will generally be a programmer error, so a crash is appropriate there, but there is the edge case where the URL could come in from the user, so we shouldn't make the assumption.

Reply:
Sure thing, can do. The question is how do you wanna handle such an error? 
The `createDirectory` endpoint can throw an error as well and I can see that you handled it by ignoring it: https://github.com/pointfreeco/swift-composable-architecture/blob/8624a8e2c35f78e5e61d5f12c69110e9020e80c8/Sources/ComposableArchitecture/SharedState/PersistenceKey/FileStorageKey.swift#L73 
Do you think it would be appropriate solution here? Or maybe we should emit error down to the user land?